pkgs/build-support/fetchpatch/default.nix

```diff
@@ -27,57 +31,63 @@ let
-          --addnewprefix=b/${lib.escapeShellArg extraPrefix} \
-      ''} \
-      --clean "$out" > "$tmpfile"
+      "${patchutils}/bin/lsdiff" \
+        ${lib.optionalString (relative != null) "-p1 -i ${lib.escapeShellArg relative}/'*'"} \
+        "$out" \
+      | sort -u | sed -e 's/[*?]/\\&/g' \
+      | xargs -I{} \
+        "${patchutils}/bin/filterdiff" \
+        --include={} \
+        --strip=${toString stripLen} \
+        ${lib.optionalString (extraPrefix != null) ''
+            --addoldprefix=a/${lib.escapeShellArg extraPrefix} \
+            --addnewprefix=b/${lib.escapeShellArg extraPrefix} \
+        ''} \
+        --clean "$out" > "$tmpfile"
 
-    if [ ! -s "$tmpfile" ]; then
-      echo "error: Normalized patch '$tmpfile' is empty (while the fetched file was not)!" 1>&2
-      echo "Did you maybe fetch a HTML representation of a patch instead of a raw patch?" 1>&2
-      echo "Fetched file was:" 1>&2
-      cat "$out" 1>&2
-      exit 1
-    fi
+      if [ ! -s "$tmpfile" ]; then
+        echo "error: Normalized patch '$tmpfile' is empty (while the fetched file was not)!" 1>&2
+        echo "Did you maybe fetch a HTML representation of a patch instead of a raw patch?" 1>&2
+        echo "Fetched file was:" 1>&2
+        cat "$out" 1>&2
+        exit 1
+      fi
 
-    ${patchutils}/bin/filterdiff \
-      -p1 \
-      ${builtins.toString (builtins.map (x: "-x ${lib.escapeShellArg x}") excludes)} \
-      ${builtins.toString (builtins.map (x: "-i ${lib.escapeShellArg x}") includes)} \
-      "$tmpfile" > "$out"
+      ${patchutils}/bin/filterdiff \
+        -p1 \
+        ${builtins.toString (builtins.map (x: "-x ${lib.escapeShellArg x}") excludes)} \
+        ${builtins.toString (builtins.map (x: "-i ${lib.escapeShellArg x}") includes)} \
+        "$tmpfile" > "$out"
 
-    if [ ! -s "$out" ]; then
-      echo "error: Filtered patch '$out' is empty (while the original patch file was not)!" 1>&2
-      echo "Check your includes and excludes." 1>&2
-      echo "Normalized patch file was:" 1>&2
-      cat "$tmpfile" 1>&2
-      exit 1
-    fi
-  '' + lib.optionalString revert ''
-    ${patchutils}/bin/interdiff "$out" /dev/null > "$tmpfile"
-    mv "$tmpfile" "$out"
-  '' + postFetch;
-} // builtins.removeAttrs args [
-  "relative" "stripLen" "extraPrefix" "excludes" "includes" "revert" "postFetch"
-])
+      if [ ! -s "$out" ]; then
+        echo "error: Filtered patch '$out' is empty (while the original patch file was not)!" 1>&2
+        echo "Check your includes and excludes." 1>&2
+        echo "Normalized patch file was:" 1>&2
+        cat "$tmpfile" 1>&2
+        exit 1
+      fi
+    '' + lib.optionalString revert ''
+      ${patchutils}/bin/interdiff "$out" /dev/null > "$tmpfile"
+      mv "$tmpfile" "$out"
+    '' + postFetch;
+  } // builtins.removeAttrs args [
+    "relative" "stripLen" "extraPrefix" "excludes" "includes" "revert" "postFetch" "reason" "throwBool" "condition" "path"
```

Comment:
We shouldn't be programming like this. (Not your fault, I know)

Reply:
do you have an ideal example in nixpkgs?

Reply:
Everything that's already separate.
Example: packages and package sets. If I create a new package `foo`, I don't have to add `remoteAttrs args ["foo"]` to `callPackage`.
Something that's not been done yet: a `mkDerivation` that doesn't try to forward _every_ argument to the builder as an environment variable / drvAttr. I've seen the structured attrs PR contain what seems to be something of a fix for this, but that seems to be too good to be true.